GearRent/Networking/APIManager.m

```diff
@@ -55,6 +64,259 @@ void fetchListingsWithCoordinates(NSArray<CLLocation *> *polygonCoordinates, voi
+                                                           NSString *dateString = [holiday valueForKey:@"date"];
+                                                           NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
+                                                           [dateFormatter setDateFormat:@"yyyy-MM-dd"];
+                                                           NSDate *date = [dateFormatter dateFromString:dateString];
+                                                           if([date isEqualToDate:today]){
+                                                               completion(YES, nil);
+                                                           }
+                                                       }
+                                                       completion(NO, nil);
+                                                   } else {
+                                                       NSLog(@"Error: %@", error);
+                                                       completion(nil, error);
+                                                   }
+                                               }
+        self->isTodayAHolidayTask = nil;
+    }];
+   [self->isTodayAHolidayTask resume];
+}
+
+void fetchDynamicPrice(Listing *listing, void(^completion)(float, NSError *)) {
+    NSInteger lookback = 24;
+    __block NSInteger numberOfSearchesForCategoryInPast24Hours = 0;
+    __block NSInteger numberOfReservationsSentInPast24Hours = 0;
+    __block float supplyFactor = 0.0;
+    __block BOOL holiday = NO;
+    dispatch_group_t serviceGroup = dispatch_group_create();
+    dispatch_group_enter(serviceGroup);
+    [APIManager searchesInPastHours:lookback categoryId:listing.categoryId location:listing.location completion:^(NSInteger result, NSError *error) {
+        if(error){
+            NSLog(@"%@", error);
+        } else{
+            numberOfSearchesForCategoryInPast24Hours = result;
+        }
+        dispatch_group_leave(serviceGroup);
+    }];
+    dispatch_group_enter(serviceGroup);
+    [APIManager reservationsInPastHours:[listing objectId] hours:lookback completion:^(NSInteger result, NSError *error) {
+        if(error){
+            NSLog(@"%@", error);
+        } else{
+            numberOfReservationsSentInPast24Hours = result;
+        }
+        dispatch_group_leave(serviceGroup);
+    }];
+    dispatch_group_enter(serviceGroup);
+    [APIManager supplyAvailableForListing:listing completion:^(float result, NSError *error) {
+        supplyFactor = result;
+        dispatch_group_leave(serviceGroup);
+    }];
+    dispatch_group_enter(serviceGroup);
+    [[APIManager alloc] isTodayAHoliday:^(BOOL isTodayAHoliday, NSError *error) {
+        if(error == nil){
+            holiday = isTodayAHoliday;
+        }
+        dispatch_group_leave(serviceGroup);
+    }];
+    dispatch_group_notify(serviceGroup, dispatch_get_main_queue(), ^{
```

Comment:
One improvement here is to create your own dispatch queue instead of the main queue(dispatch_get_main_queue()) so that the code below won't block the UI work in the main thread. 

To do that, you could create a `dispatch_queue_t _queue` ivar(instance variable), in the initializer   `_queue = dispatch_queue_create("com.gearrent.your_queue_name", DISPATCH_QUEUE_SERIAL);` Since it's not on the main queue and UI updates only happen on the main queue, the UI updates in the completion block need to be wrapped to dispatch to the main queue. 
`
      dispatch_async(
        dispatch_get_main_queue(),
        ^{
          // Your UI update e.g. self.locationLabel.text = xxx;
        }
      );
`

Reply:
Why does the current queue I have block the main thread? I thought the whole point of doing this was so that the requests were async and not blocking the UI thread

Reply:
> Why does the current queue I have block the main thread?

Since you used `dispatch_get_main_queue()` here. The async/sync and main/background thread are two different concepts. 

You can perform async tasks on main thread(like you do here) which won't blocking this line from continuing. However, the tasks(between 182 & 191)  are still scheduled to run on the main thread after the prev tasks on `serviceGroup` are completed which could affect UI when it's run(depending on how heavy-lifting the task is). The alternative is to schedule the  tasks(between 182 & 191) on a serial queue which'll run on background threads.